test-common/utils.c

```diff
@@ -579,14 +583,201 @@ maybe_hashtable_lookup (GHashTable *table,
+{
+  g_clear_pointer (&info->collection_ref, ostree_collection_ref_free);
+
+  g_free (info);
+}
+
+EosTestUpdaterCommitGraph *
+eos_test_updater_commit_graph_new (void)
+{
+  EosTestUpdaterCommitGraph *commit_graph = g_new0 (EosTestUpdaterCommitGraph, 1);
+
+  commit_graph->commit_infos = g_hash_table_new_full (g_direct_hash,
+                                                      g_direct_equal,
+                                                      NULL,
+                                                      (GDestroyNotify) eos_test_updater_commit_info_free);
+  commit_graph->parents = g_hash_table_new (g_direct_hash, g_direct_equal);
+  commit_graph->ref_cnt = 1;
+
+  return commit_graph;
+}
+
+EosTestUpdaterCommitGraph *
+eos_test_updater_commit_graph_ref (EosTestUpdaterCommitGraph *commit_graph)
+{
+  ++commit_graph->ref_cnt;
+
+  return commit_graph;
+}
+
+void
+eos_test_updater_commit_graph_unref (EosTestUpdaterCommitGraph *commit_graph)
+{
+  g_assert (commit_graph->ref_cnt > 0);
+
+  if (--commit_graph->ref_cnt != 0)
+    return;
+
+  g_clear_pointer (&commit_graph->commit_infos, g_hash_table_unref);
+  g_clear_pointer (&commit_graph->parents, g_hash_table_unref);
+
+  g_free (commit_graph);
+}
+
+typedef gpointer (*CopyFunc) (gpointer);
+
+static gpointer
+no_copy (gpointer value)
+{
+  return value;
+}
+
+/* Flip keys to values and vice versa */
+static GHashTable *
+reverse_hashtable (GHashTable     *ht,
+                   GHashFunc       hash_func,
+                   GEqualFunc      equal_func,
+                   GDestroyNotify  key_destroy,
+                   GDestroyNotify  value_destroy,
+                   CopyFunc        key_copy,
+                   CopyFunc        value_copy)
+{
+  g_autoptr(GHashTable) reversed_ht = g_hash_table_new_full (hash_func,
+                                                             equal_func,
+                                                             key_destroy,
+                                                             value_destroy);
+  gpointer key, value;
+  GHashTableIter iter;
+
+  g_hash_table_iter_init (&iter, ht);
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    g_hash_table_insert (reversed_ht,
+                         (*key_copy) (value),
+                         (*value_copy) (key));
+
+  return g_steal_pointer (&reversed_ht);
+}
+
+void
+eos_test_updater_insert_commit_steal_info (EosTestUpdaterCommitGraph *commit_graph,
+                                           guint                      parent_commit,
+                                           EosTestUpdaterCommitInfo  *commit_info)
+{
+  g_hash_table_insert (commit_graph->commit_infos,
+                       GUINT_TO_POINTER (commit_info->sequence_number),
+                       commit_info);
+  g_hash_table_insert (commit_graph->parents,
+                       GUINT_TO_POINTER (commit_info->sequence_number),
+                       GUINT_TO_POINTER (parent_commit));
```

Comment:
Why not combine `commit_graph->commit_infos` and `commit_graph->parents` by adding a `parent` field to the `EosTestUpdaterCommitInfo` struct?

Reply:
Could do.